app/src/cc/arduino/packages/formatter/AStyle.java

```diff
@@ -76,15 +80,64 @@ public void init(Editor editor) {
 
   @Override
   public void run() {
-    String originalText = editor.getCurrentTab().getText();
-    String formattedText = aStyleInterface.AStyleMain(originalText, formatterConfiguration);
 
-    if (formattedText.equals(originalText)) {
-      editor.statusNotice(tr("No changes necessary for Auto Format."));
-      return;
-    }
+    SketchTextArea textArea = editor.getCurrentTab().getTextArea();
+
+    String originalText = textArea.getSelectedText();
+
+    // If no selection use all file
+    if (originalText == null || originalText.isEmpty()) {
+
+      String formattedText = aStyleInterface.AStyleMain(textArea.getText(), formatterConfiguration);
+      editor.getCurrentTab().setText(formattedText);
+
+    } else {
+      try {
+
+        // apply indentation control keywords.
+        String FORMAT_ON = "\n// *INDENT-ON* DYN\n";
+        String FORMAT_OFF = "\n// *INDENT-OFF* DYN\n";
 
-    editor.getCurrentTab().setText(formattedText);
+        RSyntaxDocument content = (RSyntaxDocument) textArea.getDocument();
+
+        textArea.beginAtomicEdit();
+
+        int selStart = editor.getCurrentTab().getSelectionStart();
+        int selEnd = editor.getCurrentTab().getSelectionStop();
+        int lineStart = textArea.getLineOfOffset(selStart);
+        int lineEnd = textArea.getLineOfOffset(selEnd);
+
+        // Calculate offsets from begin and end of each line.
+        int fristLineOffset = textArea.getLineStartOffset(lineStart);
+        int lastLineOffset = textArea.getLineEndOffset(lineEnd);
```

Comment:
Does this line offset stuff also work correctly when the selection boundaries are just before or just after the newline? Probably something to add a test for all four cases.

Reply:
Is this the example where I need to format 1 line only, and the cursor is at the beginning of the next line? It will force formatting on the 2 lines

Reply:
I think so, yes. I think I would expect that if I select no characters on the second line, just the newline separating both, I would only format the first line, not the second?

For the start of the selection, it's probably mirrored: If I select just the preceding newline and no characters on the previous line, I would expect to not format the previous line.